android/app/src/main/java/com/sample/MainActivity.kt

```diff
@@ -32,21 +35,67 @@ class MainActivity : ReactActivity() {
         super.onCreate(savedInstanceState)
         setContentView(R.layout.activity_main)
 
+        val reactNativeHost = reactNativeHost
+        if (reactNativeHost is TestAppReactNativeHost) {
+            reactNativeHost.currentActivity = this
+        }
+
         val manifest = manifestProvider.manifest
-            ?: throw IllegalStateException("app.json is not provided or TestApp is misconfigured")
+                ?: throw IllegalStateException("app.json is not provided or TestApp is misconfigured")
+        findViewById<MaterialToolbar>(R.id.topAppBar).apply {
+            title = manifest.displayName
+            setOnMenuItemClickListener { menuItem ->
+                when (menuItem.itemId) {
+                    R.id.loadEmbeddedJsBundle -> {
+                        reload(this, true)
+                        true
+                    }
+                    R.id.loadFromDevServer -> {
+                        reload(this, false)
+                        true
+                    }
+                    R.id.showDevOptions -> {
+                        reactInstanceManager.devSupportManager.showDevOptionsDialog()
+                        true
+                    }
+                    else -> false
+                }
+            }
+        }
+
         val components = manifest.components.map {
             ComponentViewModel(it.key, it.value.displayName ?: it.key)
         }
-
-        supportActionBar?.title = manifest.displayName
-
         findViewById<RecyclerView>(R.id.recyclerview).apply {
             layoutManager = LinearLayoutManager(context)
             adapter = ComponentListAdapter(
-                LayoutInflater.from(context), components, listener
+                    LayoutInflater.from(context), components, listener
             )
 
             addItemDecoration(DividerItemDecoration(context, DividerItemDecoration.VERTICAL))
         }
     }
+
+    private fun reload(toolbar: MaterialToolbar, useEmbeddedBundle: Boolean) {
```

Comment:
Nit: extract this method into `TestAppReactNativeHost`, except the piece responsible for the Toolbar? This way we can re-use from other parts of the app if needed + simplifies this activity.